### ISSUE TITLE
elfeed-meta not always setf'able?

### ISSUE THREAD

**Issue by alphapapa** [skeeto/elfeed#292]
Hi Chris,

I'm guessing that I'm doing something wrong, but anyway, I noticed some errors in my Elfeed log buffer today, like:

`(void-function (elfeed-meta entry :title))`

That may not be verbatim, as I foolishly killed the log buffer, and now I can't reproduce it until there is new content in the feeds that I made new-entry taggers for.

Anyway, here's an example:
```el
(defun ap/elfeed/reddit.com/user--rewrite-title (entry)
    (pcase-let* ((user (elfeed-meta entry :author))
                 (title (second (s-match (rx-to-string `(seq (eval user) (1+ space) "on" (1+ space) (group (1+ anything))))
                                         (elfeed-entry-title entry))))
                 (sub (car (elfeed-meta entry :categories)))
                 (new-title (format$ "on /r/$sub/$title")))
      (unless (s-starts-with? "on /r/" title)
        ;; Doesn't already have prefix
        (setf (elfeed-meta entry :title) new-title))))
```
That function rewrites the title of entries from Reddit user comment feeds, like https://www.reddit.com/user/skeeto/.rss, adding the name of the subreddit it was posted on.

I've been using it for some time, but it seems to have stopped working recently; I'm guessing it was when I upgraded to Emacs 26, but I can't say for certain.

Today I noticed the errors in the log.  So I tried changing the `setf` to `plist-put` using `elfeed-entry-meta`:
```el
(defun ap/elfeed/reddit.com/user--rewrite-title (entry)
    (pcase-let* ((user (elfeed-meta entry :author))
                 (title (second (s-match (rx-to-string `(seq (eval user) (1+ space) "on" (1+ space) (group (1+ anything))))
                                         (elfeed-entry-title entry))))
                 (sub (car (elfeed-meta entry :categories)))
                 (new-title (format$ "on /r/$sub/$title")))
      (unless (s-starts-with? "on /r/" title)
        ;; Doesn't already have prefix
        (plist-put (elfeed-entry-meta entry) :title new-title))))
```
After refreshing feeds, that seems to have fixed the error and made the function correctly modify titles again.

But I don't understand why this is necessary.  It used to work, and I can't find any commits in the repo that seem relevant.  The Elfeed tests even use `(setf (elfeed-meta entry ...`, but apparently not with the `:title` key, so I wonder if there's something going on there.

Thanks for your help.

**Comment by skeeto** [comment]
This sounds like the sort of issue that arises if you update Emacs but 
continue using your original elpa/ directory. Byte code is backwards 
compatible, but macros are not. Some macro expansions, including setf 
expansions, may be incorrect for the new release. (This also 
occasionally happens updating packages in place, though I don't think 
that's the problem here.)

**Comment by alphapapa** [comment]
Ah, yes, of course.  Thanks.  I'll investigate and reopen if necessary.

**Comment by alphapapa** [comment]
Hi Chris,

I'm sorry to bother you about this again, because it's probably me doing something wrong, but I'm still having this problem, and I don't know why.  Here's what happens:

1.  I start Emacs.
2.  Run `elfeed`.
3.  Run `elfeed-search-fetch`.
4.  The fetch works properly, but some taggers are not applied to new entries.  In the elfeed-log buffer, I see entries like `[2018-09-06 22:28:31] [error]: https://lobste.rs/rss: (void-function (setf elfeed-entry-link))`.
5.  I put point on an entry to which that tagger applies, and manually apply the tagger by eval'ing:

```el
(with-current-buffer "*elfeed-search*"
  (let ((entry (car (elfeed-search-selected))))
    (ap/elfeed/lobste.rs--rewrite-link entry)))
```

 This fails with the same `(void-function (setf elfeed-entry-link))` error.

6.  I re-evaluate the function which calls `(setf (elfeed-entry-link ...`:
```el
(defun ap/elfeed/lobste.rs--rewrite-link (entry)
  "Replace link in lobste.rs entry with link to comments page."
  (-let (((namespace . comments-link) (elfeed-entry-id entry)))
    (when (and comments-link
               (s-prefix-p "http" comments-link))
      (setf (elfeed-entry-link entry) comments-link))))
```
7.  I manually apply the tagger again, as in step 5.
8.  The function sets and returns the value correctly.
9.  Afterward, the tagger works correctly when calling `elfeed-search-fetch`.

I'm using `use-package` for my elfeed config, which looks like this (removing everything not relevant):
```el
(use-package elfeed
  :config
  (add-hook 'elfeed-new-entry-hook
            (elfeed-make-tagger :feed-url "\\`\\(https?\\)://lobste\\.rs"
                                :callback (defun ap/elfeed/lobste.rs--rewrite-link (entry)
                                            "Replace link in lobste.rs entry with link to comments page."
                                            (-let (((namespace . comments-link) (elfeed-entry-id entry)))
                                              (when (and comments-link
                                                         (s-prefix-p "http" comments-link))
                                                (setf (elfeed-entry-link entry) comments-link)))))))
```
The expansion looks like this (expanding the original form, removing everything not relevant):
```el
(progn
  (defvar use-package--warning97
    #'(lambda (keyword err)
        (let ((msg (format "%s/%s: %s" 'elfeed keyword (error-message-string err))))
          (display-warning 'use-package msg :error))))
  (condition-case-unless-debug err
      (if (not (require 'elfeed nil t))
          (display-warning 'use-package (format "Cannot load %s" 'elfeed) :error)
        (condition-case-unless-debug err
            (progn
              (add-hook 'elfeed-new-entry-hook
                        (elfeed-make-tagger :feed-url "\\`\\(https?\\)://lobste\\.rs"
                                            :callback
                                            (defun ap/elfeed/lobste\.rs--rewrite-link (entry)
                                              "Replace link in lobste.rs entry with link to comments page."
                                              (-let (((namespace . comments-link) (elfeed-entry-id entry)))
                                                (when (and comments-link
                                                           (s-prefix-p "http" comments-link))
                                                  (setf (elfeed-entry-link entry) comments-link))))))
              t)
          (error (funcall use-package--warning97 :config err))))
    (error (funcall use-package--warning97 :catch err))))
```
So it's doing `(eval-after-load 'elfeed ...`, which should run after `elfeed` loads, which should load `elfeed-db`, which is where `elfeed-entry-link` is defined.  (Putting the `defun` directly in the `:callback` argument does seem unorthodox, but it appears to work.  And I have this same problem with other taggers whose callbacks are defined outside the `elfeed-make-tagger` call, so I don't think that's the problem.)

So as far as I can tell, the function that calls `(setf (elfeed-entry-link ...` should be defined after `elfeed-entry-link` is defined.  In fact, before I re-eval'ed that function, I did `M-x describe-function RET elfeed-entry-link RET`, and the help buffer showed this:
```
elfeed-entry-link is a compiled Lisp function in
‘~/.homesick/repos/emacs.d/emacs.d/elpa/elfeed-20180713.1229/elfeed-db.el’.

(elfeed-entry-link CL-X)

This function has a compiler macro ‘elfeed-entry-link--cmacro’.

Access slot "link" of ‘(elfeed-entry (:constructor elfeed-entry--create))’ struct CL-X.
```

Also, I deleted and reinstalled the `elfeed` package, but it didn't help.

Again, this is probably my doing something wrong rather than a bug in Elfeed, so if you don't have time to help with this, feel free to politely tell me off, haha.  :)  Thanks.

**Comment by leafpile** [comment]
I'm having the same problem with `(setf (elfeed-entry-title entry) mystr)`.  I put these definitions after `:config` in the `use-package` setup (which should be evaluated only after loading `elfeed`), but it seems that I always need to re-evaluate the functions definitions involving `setf` for them to work.  Is there any quick fix?  Thanks!

**Comment by alphapapa** [comment]
Hi Chris,

Thanks for your patience.  I did some more digging:

* There was a byte-compiled version of the file in question, but it was very old, and I have `load-prefer-newer` enabled, so it was not being used.  I also know that it was not being used because 1) changes I made to the source file were in effect after restarting Emacs, and 2) I deleted the `.elc` file, and this problem remained.  So in this case, the problem is not being caused by byte-compilation.
* I added `(eval-when-compile (require 'elfeed))` immediately before the tagger functions are defined, and this seems to have fixed the problem!
* However, I don't understand why this fixes it.  I looked at the `use-package` expansion again, and the tagger functions are defined in a `(eval-after-load 'elfeed` block.  But I guess it's probably either a bug in `use-package` or a mistake in the way I'm using it.  It can't be a bug in Elfeed.

Thanks very much for your help, as always.

@leafpile Try the solution I found; I bet it will work for you too.
